### DISCUSSION TITLE
Berry scripting: Using 'assert' with asynchronous processes

### DISCUSSION THREAD

**Discussion by wjohn007** [arendst/Tasmota#19636] [Q&A]
Following example:
```
timerName="myTimer"
timeoutDone = false

def TimeOutHandler()
    print ("start TimeOutHandler")
    timeoutDone=true
end

tasmota.set_timer(3000, TimeOutHandler,timerName);

# timer should have expired
waitMax = tasmota.millis() + 6000
while tasmota.millis() < waitMax
    tasmota.yield()
    tasmota.delay(1)
end

# 'assert' every time fails here,  the timer is not called back within the while loop
assert(timeoutDone, "should be true here")
```
The timer is not called back within the while loop. 
Is there any solution ?

John

**Comment by sfromis** [comment]
The issue is that your expectation of asynchronous processes is not available, and your loop to waste time is not a valid programming technique in the Tasmota environment for the embedded language Berry. Any loop you have blocks Tasmota totally until you exit, meaning after the final assert. At that time, the Tasmota main loop will run again, and the scheduled timer will be checked.

Anything you do in Berry should leave the code again quickly, and not have loops to manage timing. Instead use features like tasmota.set_timer or cron to handle timing. Another way is driver callbacks `every_200ms` (and others) for repeated calls, or maybe `tasmota.add_fast_loop` for a function called every 5 ms.

**Comment by wjohn007** [comment]
@sfromis 
Thank you for your quick answer.

"_Anything you do in Berry should leave the code again quickly, and not have loops to manage timing._"

I completely agree with you, and that's how I practice it for productive code too.

The question is, how can I create a unit-test that is supposed to check a process that has time delays.

Pseudocode
```
Start the process with a time delay
Wait for some time for the process to complete
Verify the results with 'assert'
```

**Reply by sfromis** [reply]
That depends on your ambition level for how to control such an unit test, what you specifically mean by "has time delays", and what types of cases to be tested, but a rudimentary strategy could start along these techniques:
```
def tester(f)
  var starttime = tasmota.millis()
  f()
  print('duration', tasmota.millis() - starttime, 'ms')
end
```

In general forget about "wait for". As long as you have a good test for "complete", another strategy could be (5 ms resolution):
```
  tasmota.add_fast_loop(
    def() 
      if completed()
        print('completed now')
      end
    end)
```
Obviously, more meat on the skeleton needed, like saving a `starttime` etc. I'd not expect this to "win over" blocking code.

Don't expect to catch a blocking function until completed, and such functions may cause the watchdog to bite, resulting in Tasmota restarting.
